### PR TITLE
feat: Run report modification callbacks before native crash reports

### PR DESCRIPTION
**Pull request by kattrali**
## Goal

Added support for modifying all reports (including native crash reports) prior to delivery.

## Design

Added new functionality to convert an Error cached on disk back into an Error object, which is then added to reports. The new callbacks are run on the resulting report object just prior to delivery.

## Changeset

|||
|--|--|
|Source| `15 files changed, 645 insertions(+), 33 deletions(-)`|
|Tests| `14 files changed, 1166 insertions(+), 7 deletions(-)` (mostly fixtures)|

### Added

* `ErrorReader` - A collection of static methods to read a report from a file, throwing an exception if the report is in some way invalid (such as broken JSON or missing required fields)
* `Configuration.beforeSend` - New callbacks are added to the configuration directly before calling `Bugsnag.init`

### Removed

* `Report.errorFile` - when constructed with a file, a report then creates an Error object, so every report now has an `Error`

## Tests

* Added new unit tests for deserializing an Error using the Reader class
* Added new integration tests for running `beforeSend` callbacks before native/and non-native crashes and notify() calls

## Discussion

This new configuration option bypasses adding convenience methods to Bugsnag/Client to avoid cases where reports may be delivered prior to the callback being added. Something like:

```java
Bugsnag.init(this); // <-- reports start sending
Bugsnag.beforeSend( ... ); // <-- already too late
```

## Review

This pull request is ready for final review, in particular:

- [ ] Performance concerns during serialization
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise?
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: sdk/src/main/java/com/bugsnag/android/Client.java [hunk 1]
```diff
@@ -1207,6 +1207,10 @@ public void disableExceptionHandler() {
     }
 
     void deliver(@NonNull Report report, @NonNull Error error) {
+        if (!runBeforeSendTasks(report)) {
+            Logger.info("Skipping notification - beforeSend task returned false");
+            return;
+        }
         try {
             config.getDelivery().deliver(report, config);
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Client.java]: Its run before every delivery attempt, as the post-callback state is not cached to disk in the event of failure.

File: sdk/src/main/java/com/bugsnag/android/ErrorStore.java [hunk 2]
```diff
@@ -133,7 +133,18 @@ private void flushReports(Collection<File> storedReports) {
 
     private void flushErrorReport(File errorFile) {
         try {
-            Report report = new Report(config.getApiKey(), errorFile);
+            Report report = new Report(config, errorFile);
+            for (BeforeSend beforeSend : config.getBeforeSendTasks()) {
+                try {
+                    if (!beforeSend.run(report)) {
+                        deleteStoredFiles(Collections.singleton(errorFile));
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/ErrorStore.java]: If the callback returned false, then the report should not be sent. This mirrors the behavior of beforeNotify.

File: sdk/src/main/java/com/bugsnag/android/Breadcrumb.java [hunk 3]
```diff
@@ -46,6 +46,16 @@
         this.name = name;
     }
 
+    Breadcrumb(@NonNull String name,
+               @NonNull BreadcrumbType type,
+               @NonNull Date captureDate,
+               @NonNull Map<String, String> metadata) {
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Breadcrumb.java]: It would be odd to require any code which creates a breadcrumb to create a date, wouldn't it? The default behavior should continue to be that a new breadcrumb is the current date.

File: sdk/src/main/java/com/bugsnag/android/ErrorReader.java [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+        try {
+            reader = new BufferedReader(new FileReader(errorFile));
+            String line;
+
+            while ((line = reader.readLine()) != null) {
+                json.append(line);
+            }
+            try {
+                JSONObject contents = new JSONObject(json.toString());
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/ErrorReader.java]: All of the contents of the file will create objects and store references to the strings in the original file, so all of the contents will be in memory for callbacks. We could reduce the incidence by only deserializing the file if there are callbacks to run though. More thoughts?

File: sdk/src/main/java/com/bugsnag/android/Report.java [hunk 5]
```diff
@@ -14,9 +14,6 @@
  */
 public class Report implements JsonStream.Streamable {
 
-    @Nullable
-    private final File errorFile;
-
     @Nullable
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Report.java]: Wouldn't that be a breaking change? Any existing reference to `report.error?` would need to remove the `?`, right?

File: sdk/src/main/java/com/bugsnag/android/HandledState.java [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
     @StringDef({REASON_UNHANDLED_EXCEPTION, REASON_STRICT_MODE, REASON_HANDLED_EXCEPTION,
         REASON_USER_SPECIFIED, REASON_CALLBACK_SPECIFIED, REASON_PROMISE_REJECTION,
-        REASON_LOG})
+        REASON_LOG, REASON_SIGNAL})
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/HandledState.java]: Yes, it can only be created using the constructor. This exception type can only be generated via the NDK handler.

File: sdk/src/main/java/com/bugsnag/android/Bugsnag.java [hunk 7]
```diff
@@ -326,12 +326,14 @@ public static void setSessionTrackingApiClient(@NonNull SessionTrackingApiClient
 
 
     /**
-     * Add a "before notify" callback, to execute code before every
-     * report to Bugsnag.
+     * Add a "before notify" callback, to execute code before sending
+     * non-fatal reports to Bugsnag.
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Bugsnag.java]: C/C++ _and_ killed the app immediately. C `notify()` uses beforeNotify callbacks.

[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Bugsnag.java]: In the paragraph below this line, the doc comment goes into more detail stating:

> "Before notify" callbacks do not run before reports generated in the event of immediate app termination, like crashes with C/C++ code.

File: sdk/src/main/java/com/bugsnag/android/Configuration.java [hunk 8]
```diff
@@ -634,6 +645,33 @@ public void setDelivery(@NonNull Delivery delivery) {
...
+     *   error.setContext(getImportantField(error));
+     *   return true;
+     * });
+     * </pre>
+     * @param beforeSend a callback to run before delivering errors to Bugsnag
+     * @see BeforeSend
+     * @see Report
+     */
+    public void beforeSend(BeforeSend beforeSend) {
+        if (!beforeSendTasks.contains(beforeSend)) {
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Configuration.java]: Good call on testing that order stays the same. I don't see why we would test that tasks are unique though. Why would it matter is someone adds the same callback multiple times?

[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/Configuration.java]: Due to lifecycle events the same callback could be unintentionally added multiple times, leading to confusion. Example: #127

File: sdk/src/main/java/com/bugsnag/android/Bugsnag.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/Bugsnag.java]: My reading of this is that `beforeNotify` is not invoked on Throwables caught by the `UncaughtExceptionHandler`, which is false. I think we should reword to avoid giving that impression.

File: sdk/src/androidTest/java/com/bugsnag/android/ErrorReaderTest.java [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+    private Error error = null;
+
+    /** Constructs an Error from a JSON file */
+    @Before
+    public void setUp() throws IOException {
+        InputStream input = InstrumentationRegistry.getContext().getResources()
+            .openRawResource(com.bugsnag.android.test.R.raw.error);
+        File fixtureFile = File.createTempFile("error", ".json");
+        OutputStream output = new FileOutputStream(fixtureFile);
```
[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/androidTest/java/com/bugsnag/android/ErrorReaderTest.java]: @kattrali this appears to have been marked as resolved but I can't see any change to the test? It should just be the case of wrapping existing streams in buffered readers/writers.

File: sdk/src/main/java/com/bugsnag/android/Breadcrumb.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/Breadcrumb.java]: It would make it easier to test the exact timestamp of breadcrumbs as we could inject whatever date we need into the constructor, and would also save an additional method. Not a big priority though.

File: sdk/src/main/java/com/bugsnag/android/Report.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/Report.java]: It would be marked as redundant but isn't a breaking change. Switching from NonNull to Nullable *would* be a breaking change however, so we'd need to be sure about this.

File: sdk/src/androidTest/java/com/bugsnag/android/ErrorReaderTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/androidTest/java/com/bugsnag/android/ErrorReaderTest.java]: It is, hadn’t pushed yet at end of day. Didn’t think anyone was going to review last night.

File: sdk/src/main/java/com/bugsnag/android/Configuration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Configuration.java]: At that point people should fix their configuration to resolve it. For example, adding uniqueness checks would not prevent an identical callback from running multiple times if its added like:

```java
Bugsnag.beforeNotify(new BeforeNotify() {
    @Override
    public boolean run(Error error) { ... }
});
```

in a method which is called multiple times. There is no duplicate references involved to verify uniqueness.

File: sdk/src/main/java/com/bugsnag/android/Bugsnag.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (snmaynard) [reply sdk/src/main/java/com/bugsnag/android/Bugsnag.java]: I agree with @fractalwrench  on this one - that could be clearer.

File: sdk/src/main/java/com/bugsnag/android/Configuration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (snmaynard) [reply sdk/src/main/java/com/bugsnag/android/Configuration.java]: If I put the same beforeNotify function in multiple times I would expect it to run multiple times, so I'm confused why we would prevent that

File: sdk/src/main/java/com/bugsnag/android/Report.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/Report.java]: Reverting this bit to only deserialize reports if there are callbacks to be run.

**Comment by kattrali** [comment]
Made a few amendments:

1. Used `CachedThread` everywhere in `ThreadState`, replacing the existing `Thread` object cache.
2. Used `JsonReader` in the `ErrorReader` to stream the file. My only concern here is that `JsonWriter is an internal fork of some distant version while `JsonReader` is pinned to the OS. The tests cover type coercion for numbers and a few other cases which were initially buggy.
3. Added more tests around callback ordering and execution.

File: features/before_send.feature [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    Scenario: Change report before native crash report delivered
+        When I run "NativeBeforeSendScenario"
+        And I configure the app to run in the "non-crashy" state
+        And I relaunch the app
+        Then I should receive a request
+        And the request payload contains a completed native report
+        And the exception "errorClass" equals "SIGSEGV"
+        And the exception "message" equals "Segmentation violation (invalid memory reference)"
+        And the event "context" equals "!important"
```
[the remark this answers is not in the document]
Comment (kattrali) [reply features/before_send.feature]: The unit test handles a different field, and either way it isn't exhaustive for the fields which can be changed.

File: sdk/src/androidTest/java/com/bugsnag/android/ErrorReaderTest.java [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+import java.util.TimeZone;
+
+@RunWith(AndroidJUnit4.class)
+@SmallTest
+public class ErrorReaderTest {
+
+    private Error error = null;
+
+    /** Constructs an Error from a JSON file */
+    @Before
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/androidTest/java/com/bugsnag/android/ErrorReaderTest.java]: Used `@BeforeClass` on setUp to run it once instead of on every test.

File: sdk/src/main/java/com/bugsnag/android/Configuration.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (fractalwrench) [reply sdk/src/main/java/com/bugsnag/android/Configuration.java]: The activity lifecycle on Android makes it easy to add the same callback multiple times, accidentally. Consider the following code, which would add the same callback every time the user altered the device's orientation:

```
class MyActivity extends Activity implements BeforeNotify {

    // this can be called by the framework multiple times for the same instance
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState); object
        bugsnag.beforeNotify(this); // the activity object implements the callback
    }
}
```

bugsnag-android currently detects this scenario and prevents the same callback being added, by checking whether the collection already contains the object. It's worth noting that if the example code was altered to use an inner class instead, bugsnag-android would not prevent the addition:
```
bugsnag.beforeNotify(new BeforeNotify() {
    ...
});
```

This is because a new object would be created each time, and the previously added callback will not have referential equality with the new callback.

What sort of scenario would it be useful for the same callback to be invoked multiple times?

File: sdk/src/androidTest/java/com/bugsnag/android/CachedThreadTest.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+        assertEquals("ando", result.getString("type"));
+        assertEquals(true, result.getBoolean("errorReportingThread"));
+
+        JSONArray frames = result.getJSONArray("stacktrace");
+        assertEquals(3, frames.length());
+        assertEquals("run_func", frames.getJSONObject(0).getString("method"));
+        assertEquals("librunner.so", frames.getJSONObject(0).getString("file"));
+        assertEquals(5038, frames.getJSONObject(0).getInt("lineNumber"));
+        assertEquals("Runner.runFunc", frames.getJSONObject(1).getString("method"));
+        assertEquals("Runner.java", frames.getJSONObject(1).getString("file"));
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/androidTest/java/com/bugsnag/android/CachedThreadTest.java]: When you say readability, do you mean its not immediately obvious what this does, or something else?

File: sdk/src/main/java/com/bugsnag/android/ErrorReader.java [hunk 13]
```diff
@@ -0,0 +1,439 @@
...
+class ErrorReader {
+
+    /**
+     * Parses an {@link Error} cached as JSON into an Error object.
+     *
+     * @throws IOException if the file cannot be parsed into a valid JSON object,
+     *                     such as if the JSON syntax is invalid or a required
+     *                     field is missing.
+     */
+    static Error readError(@NonNull Configuration config, @NonNull File errorFile)
```
[the remark this answers is not in the document]
Comment (kattrali) [reply sdk/src/main/java/com/bugsnag/android/ErrorReader.java]: The benefit of it being static is that there isn't any state (shared or otherwise) to be mocked. Its obvious what it does and that all related logic/state is within the method itself.
Adding additional state is something we want to avoid as long as possible, so there's no need to make it easier before we actually need it.

**Comment by kattrali** [comment]
The `Reader` was vendored because Back In The Day™, bugsnag-android supported API levels < 11, so `Json*` wasn't in the standard library. Given that, I don't think we need to vendor anything new there.
